[SPARK-30234][SQL] ADD FILE cannot add directories from sql CLI

iRakson opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
Now users can add directories from sql CLI as well using ADD FILE command and setting spark.sql.addDirectory.recursive to true.
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
In SPARK-4687, support was added for adding directories as resources. But sql users cannot use that feature from CLI. 

`ADD FILE /path/to/folder` gives the following error:
`org.apache.spark.SparkException: Added file /path/to/folder is a directory and recursive is not turned on.`

Users need to turn on `recursive` for adding directories. Thus a configuration was required which will allow users to turn on `recursive`. 
Also Hive allow users to add directories from their shell.
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce any user-facing change?
Yes. Users can set recursive using `spark.sql.addDirectory.recursive`.
<!--
```
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->


### How was this patch tested?
Manually.
Will add test cases soon.
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
 SPARK SCREENSHOTS
When `spark.sql.addDirectory.recursive` is not turned on.
![Screenshot from 2019-12-13 08-02-13](https://user-images.githubusercontent.com/15366835/70765124-c6b4a100-1d7f-11ea-9352-9c010af5b38b.png)

After setting `spark.sql.addDirectory.recursive` to true.

![Screenshot from 2019-12-13 08-02-59](https://user-images.githubusercontent.com/15366835/70765118-be5c6600-1d7f-11ea-9faf-0b1c46ee299b.png)

HIVE SCREENSHOT

![Screenshot from 2019-12-13 14-44-41](https://user-images.githubusercontent.com/15366835/70788979-17e08700-1db8-11ea-9c0c-b6d6f6e80a35.png)

`RELEASE_NOTES.txt` is text file while `dummy` is a directory.

iRakson committed:
[SPARK-30234]ADD FILE cannot add directories from sql CLI

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

iRakson commented:
cc @cloud-fan @maropu

cloud-fan commented:
who else supports this feature? I'm wondering if we should introduce `ADD DIR`

iRakson commented:
As of now i have checked for hive only. It supported there using ADD FILE command only. I guess allowing to add directory will be of great use.

maropu commented:
cc: @dongjoon-hyun @wangyum cuz they are familiar with hive.

> Also Hive allow users to add directories from their shell.

Hive enables this features by default?

iRakson commented:
Yes, hive enables this feature by default.

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2098,6 +2098,13 @@ object SQLConf {
+
+  val ADD_DIRECTORY_USING_RECURSIVE =
+    buildConf("spark.sql.addDirectory.recursive")
+    .doc(" When true, users can add directory by passing path of a directory to ADD FILE" +
```
nit: leading space can be removed

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala:
```diff
@@ -24,6 +24,7 @@ import org.apache.hadoop.fs.Path
 
 import org.apache.spark.sql.{Row, SparkSession}
 import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference}
+import org.apache.spark.sql.internal.SQLConf
```
Can be removed.

iRakson replied:
Yeahh. I forgot to remove this one. I Will make all the necessary changes.

HyukjinKwon commented:
ok test

HyukjinKwon commented:
Can you show how you manually tested in PR description in both Spark and Hive?

iRakson commented:
> Can you show how you manually tested in PR description in both Spark and Hive?

I have attached the screenshot for spark. Will attach hive one a bit later.

iRakson commented:
cc @HyukjinKwon @dongjoon-hyun @maropu

iRakson committed:
[SPARK-30234]Fix Review Comments

maropu commented:
Can you add tests, too?

iRakson commented:
Yeah. I will add them in a bit.

iRakson committed:
[SPARK-30234] Added TestSuite

iRakson commented:
Test Suite added. @maropu @cloud-fan @HyukjinKwon

iRakson commented:
Can you review the test suite.
cc @maropu @cloud-fan @HyukjinKwon

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2098,6 +2098,13 @@ object SQLConf {
+
+  val ADD_DIRECTORY_USING_RECURSIVE =
+    buildConf("spark.sql.addDirectory.recursive")
+      .doc(" When true, users can add directory by passing path of a directory to ADD FILE" +
```
nit: drop a space in the head.

maropu replied:
Also, it seems we have a space in the end.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/ResourcesSuite.scala:
```diff
@@ -0,0 +1,49 @@
+      val dirPath = dir.getAbsolutePath
+        intercept[SparkException] {
+          sparkSession.sql(s"ADD FILE $dir")
+        }.getMessage.contains(s" Added file $dirPath is a directory and recursive is not turned on")
```
plz use `assert`.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/ResourcesSuite.scala:
```diff
@@ -0,0 +1,49 @@
+      sparkSession.sql(s"ADD FILE $testDir")
+      val sep = File.separator
+      if(!new File(SparkFiles.get(testDir.getName + sep + testFile1.getName)).exists()) {
+        throw new SparkException("TestFile1 Not found.")
```
plz use `assert`here, too.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/ResourcesSuite.scala:
```diff
@@ -0,0 +1,49 @@
+  test("Add Directory when ADD_DIRECTORY_USING_RECURSIVE set to true") {
+    withTempDir { testDir =>
+      val testFile1 = File.createTempFile("testFile", "1", testDir)
+      sparkSession.sql("set spark.sql.addDirectory.recursive=true")
```
plz use `withSQLConf`.

iRakson replied:
I just used this to simulate spark-sql commands. 
I will use` withSQLConf` now.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/ResourcesSuite.scala:
```diff
@@ -0,0 +1,49 @@
+import org.apache.spark.{SparkException, SparkFiles, SparkFunSuite}
+import org.apache.spark.sql.SparkSession
+
+class ResourcesSuite extends SparkFunSuite{
```
nit: `SparkFunSuite {`

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/ResourcesSuite.scala:
```diff
@@ -0,0 +1,49 @@
+/*
```
How about naming `RunnableCommandSuite` instead of `ResourcesSuite`?

iRakson replied:
Yeah. That's a better. I will rename the file.

maropu replied:
Or, it might be better to rename the existing `DDLSuite` to RunnnableCommandSuite` then add new tests there instead of creating a new test file. cc: @dongjoon-hyun

iRakson replied:
I have renamed it to `RunnableCommandSuite` for the time being. 
@dongjoon-hyun Should i move the test to `DDLSuite` and rename it? or is it fine now only?

iRakson committed:
Fix Review Comments

iRakson commented:
@maropu Anything other changes needs to be done for this one?

iRakson commented:
@maropu Any other changes required for this one???

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/RunnableCommandSuite.scala:
```diff
@@ -0,0 +1,52 @@
+class RunnableCommandSuite extends SQLTestUtils {
+
+  override protected def spark: SparkSession =
+    SparkSession.builder().master("local").appName("test_session").getOrCreate()
```
You can use `SharedSparkSession`.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/RunnableCommandSuite.scala:
```diff
@@ -0,0 +1,52 @@
+          spark.sql(s"ADD FILE $dir")
+        }.getMessage
+      assert(msg.contains("is a directory and recursive is not turned on"))
+
```
nit: remove the blank line.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/RunnableCommandSuite.scala:
```diff
@@ -0,0 +1,52 @@
+
+  test("Add Directory when ADD_DIRECTORY_USING_RECURSIVE set to true") {
+    withTempDir { testDir =>
+      val testFile1 = File.createTempFile("testFile", "1", testDir)
```
nit: You can use `Utils.createTempDir` instead.

maropu replied:
nit: `testFile1` -> `testFile`

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/RunnableCommandSuite.scala:
```diff
@@ -0,0 +1,52 @@
+      withSQLConf(SQLConf.ADD_DIRECTORY_USING_RECURSIVE.key -> "true") {
+        spark.sql(s"ADD FILE $testDir")
+        val sep = File.separator
+        assert(new File(SparkFiles.get(testDir.getName + sep + testFile1.getName)).exists())
```
`assert(new File(SparkFiles.get(s"${testDir.getName}/${testFile1.getName}")).exists())`?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2098,6 +2098,12 @@ object SQLConf {
+    .doc(" When true, users can add directory by passing path of a directory to ADD FILE" +
+      "command of SQL. If false, then only a single file can be added.")
+    .booleanConf
+    .createWithDefault(false)
```
Does Hive have an option for this feature for turning off/on it?

iRakson replied:
No. There is no such option in hive. It is enabled by default.

maropu replied:
If so, how about following that?; just adding an option under `spark.sql.legacy.xxx` then enabling it by default. Also, we need to update the migration guide, too.

iRakson replied:
@maropu Do we need to add this under `spark.sql.legacy.xxx` ? Or is it fine this way only??
@HyukjinKwon @dongjoon-hyun

iRakson committed:
Fix

iRakson commented:
@maropu @HyukjinKwon  please review the changes.

iRakson committed:
Review Comment Fix

iRakson commented:
Added these changes under `spark.sql.legacy.xxx`. Also made it true by default to follow the hive functionality. 
Migration guide has been updated too.

iRakson commented:
Added these changes under spark.sql.legacy.xxx. Also made it true by default to follow the hive functionality.
Migration guide has been updated too.

@maropu @cloud-fan @HyukjinKwon Please verify the changes once.

maropu commented:
All the reviews have been addressed? 
 https://github.com/apache/spark/pull/26863/files#r357460250

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2098,6 +2098,12 @@ object SQLConf {
+
+  val LEGACY_ADD_DIRECTORY_USING_RECURSIVE = buildConf("spark.sql.legacy.addDirectory.recursive")
+    .doc(" When true, users can add directory by passing path of a directory to ADD FILE" +
+      "command of SQL. If false, then only a single file can be added.")
```
nit: You need a single space in the head.

iRakson replied:
Added a space in last line.

iRakson replied:
@maropu Any more changes required for this one??

iRakson committed:
Review Comment Fix

iRakson commented:
@cloud-fan Anything else needs to be done for this one??

iRakson commented:
@cloud-fan Please review this one as well. :)

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/RunnableCommandSuite.scala:
```diff
@@ -0,0 +1,46 @@
+
+class RunnableCommandSuite extends SQLTestUtils with SharedSparkSession {
+
+  test("Add Directory when LEGACY_ADD_DIRECTORY_USING_RECURSIVE not set to true") {
```
Yea, just move to `DDLSuite`.

`LEGACY_ADD_DIRECTORY_USING_RECURSIVE` -> actual configuration name or string iterpolation via `$LEGACY_ADD_DIRECTORY_USING_RECURSIVE`

HyukjinKwon replied:
Add Directory -> Add a directory

HyukjinKwon commented:
Looks fine otherwise. @iRakson can you resolve conflicts and update?

iRakson committed:
Fix

iRakson commented:
> Looks fine otherwise. @iRakson can you resolve conflicts and update?

Done.

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:
```diff
@@ -2909,4 +2910,23 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
     }
   }
+
+  test("Add a Directory when spark.sql.legacy.addDirectory.recursive not set to true") {
```
`Directory` -> `directory`

iRakson committed:
Fix

maropu commented:
ok to test

maropu reviewed (approved):
LGTM. Pending Jenkins. Thanks for the check, Hyukjin!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21059/
Test PASSed.

SparkQA commented:
**[Test build #116267 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116267/testReport)** for PR 26863 at commit [`837cc60`](https://github.com/apache/spark/commit/837cc604b6a68d99ea295a69a4aeeca3ca64283b).

SparkQA commented:
**[Test build #116267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116267/testReport)** for PR 26863 at commit [`837cc60`](https://github.com/apache/spark/commit/837cc604b6a68d99ea295a69a4aeeca3ca64283b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116267/
Test FAILed.

maropu commented:
retest this please

SparkQA commented:
**[Test build #116272 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116272/testReport)** for PR 26863 at commit [`837cc60`](https://github.com/apache/spark/commit/837cc604b6a68d99ea295a69a4aeeca3ca64283b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21063/
Test PASSed.

SparkQA commented:
**[Test build #116272 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116272/testReport)** for PR 26863 at commit [`837cc60`](https://github.com/apache/spark/commit/837cc604b6a68d99ea295a69a4aeeca3ca64283b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116272/
Test FAILed.

maropu commented:
retest this please

SparkQA commented:
**[Test build #116275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116275/testReport)** for PR 26863 at commit [`837cc60`](https://github.com/apache/spark/commit/837cc604b6a68d99ea295a69a4aeeca3ca64283b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21067/
Test PASSed.

SparkQA commented:
**[Test build #116275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116275/testReport)** for PR 26863 at commit [`837cc60`](https://github.com/apache/spark/commit/837cc604b6a68d99ea295a69a4aeeca3ca64283b).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116275/
Test FAILed.

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21076/
Test PASSed.

SparkQA commented:
**[Test build #116285 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116285/testReport)** for PR 26863 at commit [`837cc60`](https://github.com/apache/spark/commit/837cc604b6a68d99ea295a69a4aeeca3ca64283b).

SparkQA commented:
**[Test build #116285 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116285/testReport)** for PR 26863 at commit [`837cc60`](https://github.com/apache/spark/commit/837cc604b6a68d99ea295a69a4aeeca3ca64283b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116285/
Test FAILed.

HyukjinKwon commented:
Seems the test fails:

```
[info] - Add a directory when spark.sql.legacy.addDirectory.recursive not set to true *** FAILED *** (66 milliseconds)
[info]   Expected exception org.apache.spark.SparkException to be thrown, but no exception was thrown (DDLSuite.scala:2916)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
[info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
[info]   at org.scalatest.Assertions.intercept(Assertions.scala:824)
[info]   at org.scalatest.Assertions.intercept$(Assertions.scala:804)
[info]   at org.scalatest.FunSuite.intercept(FunSuite.scala:1560)
[info]   at org.apache.spark.sql.execution.command.DDLSuite.$anonfun$new$412(DDLSuite.scala:2916)
[info]   at org.apache.spark.sql.execution.command.DDLSuite.$anonfun$new$412$adapted(DDLSuite.scala:2915)
[info]   at org.apache.spark.sql.test.SQLTestUtils.$anonfun$withTempDir$1(SQLTestUtils.scala:76)
[info]   at org.apache.spark.sql.test.SQLTestUtils.$anonfun$withTempDir$1$adapted(SQLTestUtils.scala:75)
[info]   at org.apache.spark.SparkFunSuite.withTempDir(SparkFunSuite.scala:161)
[info]   at org.apache.spark.sql.execution.command.DDLSuite.org$apache$spark$sql$test$SQLTestUtils$$super$withTempDir(DDLSuite.scala:228)
[info]   at org.apache.spark.sql.test.SQLTestUtils.withTempDir(SQLTestUtils.scala:75)
[info]   at org.apache.spark.sql.test.SQLTestUtils.withTempDir$(SQLTestUtils.scala:74)
[info]   at org.apache.spark.sql.execution.command.DDLSuite.withTempDir(DDLSuite.scala:228)
```

and it causes the cascaded test failures.

iRakson commented:
> Seems the test fails:
> 
> ```
> [info] - Add a directory when spark.sql.legacy.addDirectory.recursive not set to true *** FAILED *** (66 milliseconds)
> [info]   Expected exception org.apache.spark.SparkException to be thrown, but no exception was thrown (DDLSuite.scala:2916)
> [info]   org.scalatest.exceptions.TestFailedException:
> [info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
> [info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
> [info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
> [info]   at org.scalatest.Assertions.intercept(Assertions.scala:824)
> [info]   at org.scalatest.Assertions.intercept$(Assertions.scala:804)
> [info]   at org.scalatest.FunSuite.intercept(FunSuite.scala:1560)
> [info]   at org.apache.spark.sql.execution.command.DDLSuite.$anonfun$new$412(DDLSuite.scala:2916)
> [info]   at org.apache.spark.sql.execution.command.DDLSuite.$anonfun$new$412$adapted(DDLSuite.scala:2915)
> [info]   at org.apache.spark.sql.test.SQLTestUtils.$anonfun$withTempDir$1(SQLTestUtils.scala:76)
> [info]   at org.apache.spark.sql.test.SQLTestUtils.$anonfun$withTempDir$1$adapted(SQLTestUtils.scala:75)
> [info]   at org.apache.spark.SparkFunSuite.withTempDir(SparkFunSuite.scala:161)
> [info]   at org.apache.spark.sql.execution.command.DDLSuite.org$apache$spark$sql$test$SQLTestUtils$$super$withTempDir(DDLSuite.scala:228)
> [info]   at org.apache.spark.sql.test.SQLTestUtils.withTempDir(SQLTestUtils.scala:75)
> [info]   at org.apache.spark.sql.test.SQLTestUtils.withTempDir$(SQLTestUtils.scala:74)
> [info]   at org.apache.spark.sql.execution.command.DDLSuite.withTempDir(DDLSuite.scala:228)
> ```
> 
> and it causes the cascaded test failures.

I will check the failing test cases.

maropu commented:
Oh, I see...

iRakson committed:
Fix

iRakson commented:
> Seems the test fails:
> 
> ```
> [info] - Add a directory when spark.sql.legacy.addDirectory.recursive not set to true *** FAILED *** (66 milliseconds)
> [info]   Expected exception org.apache.spark.SparkException to be thrown, but no exception was thrown (DDLSuite.scala:2916)
> [info]   org.scalatest.exceptions.TestFailedException:
> [info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
> [info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
> [info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
> [info]   at org.scalatest.Assertions.intercept(Assertions.scala:824)
> [info]   at org.scalatest.Assertions.intercept$(Assertions.scala:804)
> [info]   at org.scalatest.FunSuite.intercept(FunSuite.scala:1560)
> [info]   at org.apache.spark.sql.execution.command.DDLSuite.$anonfun$new$412(DDLSuite.scala:2916)
> [info]   at org.apache.spark.sql.execution.command.DDLSuite.$anonfun$new$412$adapted(DDLSuite.scala:2915)
> [info]   at org.apache.spark.sql.test.SQLTestUtils.$anonfun$withTempDir$1(SQLTestUtils.scala:76)
> [info]   at org.apache.spark.sql.test.SQLTestUtils.$anonfun$withTempDir$1$adapted(SQLTestUtils.scala:75)
> [info]   at org.apache.spark.SparkFunSuite.withTempDir(SparkFunSuite.scala:161)
> [info]   at org.apache.spark.sql.execution.command.DDLSuite.org$apache$spark$sql$test$SQLTestUtils$$super$withTempDir(DDLSuite.scala:228)
> [info]   at org.apache.spark.sql.test.SQLTestUtils.withTempDir(SQLTestUtils.scala:75)
> [info]   at org.apache.spark.sql.test.SQLTestUtils.withTempDir$(SQLTestUtils.scala:74)
> [info]   at org.apache.spark.sql.execution.command.DDLSuite.withTempDir(DDLSuite.scala:228)
> ```
> 
> and it causes the cascaded test failures.

Initially `spark.sql.legacy.addDirectory.recursive` was `false` by default. Later i changed the default value to `true` but forgot to update the test cases.

I have fixed that now. It should be fine now.

iRakson commented:
retest this please

SparkQA commented:
**[Test build #116301 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116301/testReport)** for PR 26863 at commit [`6ba556b`](https://github.com/apache/spark/commit/6ba556bb34ea1505b44c175443e9b2fe5566c02d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21092/
Test PASSed.

SparkQA commented:
**[Test build #116301 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116301/testReport)** for PR 26863 at commit [`6ba556b`](https://github.com/apache/spark/commit/6ba556bb34ea1505b44c175443e9b2fe5566c02d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116301/
Test FAILed.

iRakson committed:
Fix

SparkQA commented:
**[Test build #116348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116348/testReport)** for PR 26863 at commit [`8c94912`](https://github.com/apache/spark/commit/8c949124980b0c0fe7a97f6c9a4cbea05deb9d8a).

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21139/
Test PASSed.

SparkQA commented:
**[Test build #116348 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116348/testReport)** for PR 26863 at commit [`8c94912`](https://github.com/apache/spark/commit/8c949124980b0c0fe7a97f6c9a4cbea05deb9d8a).
 * This patch **fails due to an unknown error code, -9**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116348/
Test FAILed.

iRakson commented:
retest this please

SparkQA commented:
**[Test build #116373 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116373/testReport)** for PR 26863 at commit [`d326962`](https://github.com/apache/spark/commit/d3269622ae61dd29df555c6227a7b6ea91c4e3a3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21164/
Test PASSed.

SparkQA commented:
**[Test build #116373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116373/testReport)** for PR 26863 at commit [`d326962`](https://github.com/apache/spark/commit/d3269622ae61dd29df555c6227a7b6ea91c4e3a3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116373/
Test FAILed.

iRakson commented:
retest this please

maropu commented:
retest this please

SparkQA commented:
**[Test build #116432 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116432/testReport)** for PR 26863 at commit [`d326962`](https://github.com/apache/spark/commit/d3269622ae61dd29df555c6227a7b6ea91c4e3a3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21220/
Test PASSed.

SparkQA commented:
**[Test build #116432 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116432/testReport)** for PR 26863 at commit [`d326962`](https://github.com/apache/spark/commit/d3269622ae61dd29df555c6227a7b6ea91c4e3a3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116432/
Test PASSed.

HyukjinKwon commented:
@iRakson can you resolve conflicts? should be good to merge then.

iRakson committed:
Fix

iRakson commented:
> @iRakson can you resolve conflicts? should be good to merge then.

Resolved all the conflicts.

SparkQA commented:
**[Test build #116475 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116475/testReport)** for PR 26863 at commit [`1029613`](https://github.com/apache/spark/commit/102961320d734e29541c9131f6138e97a1e044b6).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21262/
Test PASSed.

SparkQA commented:
**[Test build #116475 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116475/testReport)** for PR 26863 at commit [`1029613`](https://github.com/apache/spark/commit/102961320d734e29541c9131f6138e97a1e044b6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116475/
Test PASSed.

HyukjinKwon commented:
Merged to master.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2147,6 +2147,12 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val LEGACY_ADD_DIRECTORY_USING_RECURSIVE = buildConf("spark.sql.legacy.addDirectory.recursive")
```
spark.sql.legacy.addDirectory.recursive.enabled?

maropu replied:
Looks ok and I'll do followup.

maropu replied:
https://github.com/apache/spark/pull/27372/files

Ngone51 replied:
Hey guys, I think the config is not consistent between name and default value.

If we want it to be a real legacy config, then, I think its name should be like `spark.sql.legacy.addSingleFileInAddFile` and default to `false`. If not, then, we may name it as `spark.sql.allowAddDirectoryInAddFile` and default to `true`.

Any ideas? @maropu @iRakson 

BTW, I'm going to revise a batch of boolean configs in order to comply with our [new config name policy](http://apache-spark-developers-list.1001551.n3.nabble.com/DISCUSS-naming-policy-of-Spark-configs-td28875.html). So, I'll address the consensus here.

HyukjinKwon replied:
Yeah, please go ahead.

maropu replied:
+1

Ngone51 replied:
So, which one do you prefer? Legacy or not legacy? Personally, I prefer `not legacy` for this new feature.

Ngone51 replied:
I'm going to pick `spark.sql.allowAddDirectoryInAddFile`. Thanks!

maropu replied:
We need to keep the old behaviour? I personally think the legacy one looks ok to me.

HyukjinKwon replied:
yup, +1 for @maropu's

iRakson replied:
I guess, it should be `spark.sql.allowAddDirectoryInAddFile` and default set to `true`. If default is set to false then we will go back to our previous behaviour which do not allow adding directory using `ADD FILE`.

Ngone51 replied:
Sorry guys, forget to tell you that this one wasn't included in that renaming PR. Because the change we need here doesn't related to the naming policy. So, I think we need to separate follow up for it.

@iRakson Can you send a follow up to rename it according to your own thoughts?

iRakson replied:
I will raise a followup for this one.

gatorsmile commented on docs/sql-migration-guide.md:
```diff
@@ -266,6 +266,8 @@ license: |
 
   - Since Spark 3.0, `location` and `comment` become reserved database properties, Commands will fail if we specify reserved properties in `CREATE DATABASE ... WITH DBPROPERTIES` and `ALTER DATABASE ... SET DBPROPERTIES`. We need their specific clauses to specify them, e.g. `CREATE DATABASE test COMMENT 'any comment' LOCATION 'some path'`. We can set `spark.sql.legacy.property.nonReserved` to `tr ...
 
+  - Since Spark 3.0, `ADD FILE` can be used to add file directories as well. Earlier only single files can be added using this command. To restore the behaviour of earlier versions, set `spark.sql.legacy.addDirectory.recursive` to false.
```
also update this line.
